src/modalities/running_env/fsdp/device_mesh.py

```diff
@@ -0,0 +1,140 @@
+from enum import Enum
+from typing import Annotated, Optional
+
+from pydantic import BaseModel, Field, model_validator
+from torch.distributed.device_mesh import DeviceMesh, init_device_mesh
+
+from modalities.exceptions import ConfigError
+from modalities.util import print_rank_0
+from modalities.utils.logging import get_logger
+
+logger = get_logger("model_factory")
+
+
+class DeviceMeshConfig(BaseModel):
+    # inspired by ParallelDims class in
+    # https://github.com/pytorch/torchtitan/blob/cfc0f4e08dc71685cdcb394464187d2eeedd1a5f/torchtitan/parallelisms/parallel_dims.py#L15
+    device_type: str = "cuda"
+    data_parallel_replicate_degree: Annotated[int, Field(strict=True, gt=0)]
+    # if -1, we will calculate the shard degree based on the world size and other parallel degrees
+    data_parallel_shard_degree: Annotated[int, Field(strict=True, ge=-1)]
+    tensor_parallel_degree: Annotated[int, Field(strict=True, gt=0)] = 1
+    pipeline_parallel_degree: Annotated[int, Field(strict=True, gt=0)] = 1
+    context_parallel_degree: Annotated[int, Field(strict=True, gt=0)] = 1
+    enable_loss_parallel: Optional[bool] = False
+    world_size: Annotated[int, Field(strict=True, gt=0)]
+
+    @model_validator(mode="after")
+    def _validate(self):
+        if (
+            self.tensor_parallel_degree != 1
+            or self.pipeline_parallel_degree != 1
+            or self.context_parallel_degree != 1
+            or self.enable_loss_parallel
+        ):
+            logger.warning(
```

Comment:
Why not raising an exception?

Reply:
I  made it a warning so we could properly test the device mesh. I will make it configurable to raise or warn.